### DISCUSSION TITLE
Dockerize the entire development environment

### DISCUSSION THREAD

**Discussion by gsong** [gsong/express-react-project-starter#38] [Ideas]
Docker filesystem performance seems to be much better now on macOS, using it for Node development definitely feels possible again.

This would eliminate the issue of people using different versions of Node and npm, or just weird environmental issues in general, but would introduce its own pitfalls such as having to be inside the container in order to add packages to the project.

**Comment by ndp** [comment]
This certainly seems like an emerging development approach (along with cloud-based dev), so it's intriguing.

I've worked with a couple students who have older machines. They seem to be right at the edge of their machine's capabilities. It's certainly possible that docker development could be lighter weight, I suspect it is not. I'd want to make sure it doesn't make it any slower for folks.

I'd also be concerned that this might would involve even more custom scripts to make things work. These are convenient, but take students away from the core tools and their documentation.

Can we specify exact versions of node and npm in package.json?

**Reply by gsong** [reply]
> Can we specify exact versions of node and npm in package.json?

As far as I know you can specify the minimum node version and that's it. Even then it's not honored. You can actually bake in the node and npm versions as dev dependencies, but that's a very heavy weight solution.

**Comment by gsong** [comment]
After more thought on this. I don't think this is worth doing as it'll deviate from standard workflows when working with Node.
